### ISSUE TITLE
Commentary gc + matchup a%

### ISSUE THREAD

**Issue by kiryph** [andymass/vim-matchup#15]
I was trying to comment a vimscript function by combining [vim-commentary's](https://github.com/tpope/vim-commentary/) operator `gc` with the text-object `a%` from match-up.

```viml
function! s:test()

    let l:str = "hello"

    return l:str

endfunction
```

Placing the cursor on `let l:str = "hello"` and pressing `gca%` changes the text to 
```viml
" function! s:test()

"     let l:str = "hello"

"     return l:str
endfunction
```
As you can see `endfunction` is not commented. Also `gci%` does not what I would expect.

Did I miss something or can this be improved?

**Comment by kiryph** [comment]
I see that the word `return` is also a target for `a%`. Using `gc1a%` works.

**Comment by kiryph** [comment]
However, given your example:
```
if condition
 █call one()
  call two()
endif
```
pressing di% will produce
```
if condition
endif
```

Should this work similar to vimscript functions?

**Comment by andymass** [comment]
Glad the first issue was resolved :)  I've debated whether this should be configurable with an option.

I'm not sure what you mean precisely regarding the second, `call` is not a target so "inner %" would be the text between the `if` and `endif` (subject to the line-wise special case, which I think you are aware of).

**Comment by andymass** [comment]
I think the solution is

   let g:matchup_text_obj_linewise_operators += [ 'g@' ]

can you confirm?

**Comment by kiryph** [comment]
No, I cannot confirm. Sorry for not adding an example for what I meant. But you already figured it out. Yes, I would expect that the `if condition` would not be commented. Consider the vimscript function from above as a second example
```vim
function! s:test()

    let l:str = "hello"

    return l:str

endfunction
```
Place the cursor on the line with `let ...` and press `gc1i%`, I get
```vim
" function! s:test()

"     let l:str = "hello"

"     return l:str

endfunction
```
I would expect
```vim
function! s:test()

"     let l:str = "hello"

"     return l:str

endfunction
```

Using `1i%` for a visual selection shows that match-up works characterwise and just omits the matchword `function`:



![screen shot 2017-12-20 at 09 18 33](https://user-images.githubusercontent.com/371756/34197541-427655fc-e567-11e7-96b7-e6ea3ec94279.png)

Well, this looks intended. I am not sure how to resolve this. Maybe enforcing linewise mode could be used, that the `i%` object does include lines with match words.

**Comment by andymass** [comment]
Strange.. I cannot reproduce it with the following vimrc:

```vim
set nocompatible

" load match-up (and commentary)
for s:loc in ['bundle', 'plugged']
    let &rtp  = '~/.vim/'.s:loc.'/vim-matchup,' . &rtp
    let &rtp .= ',~/.vim/'.s:loc.'/vim-matchup/after'
    let &rtp  = '~/.vim/'.s:loc.'/vim-commentary,' . &rtp
    let &rtp .= ',~/.vim/'.s:loc.'/vim-commentary/after'
endfor

filetype plugin indent on
syntax enable

" match-up options go here
let g:matchup_text_obj_linewise_operators = ['d', 'y', 'g@', ':']
```

However, I think using `:` for visual mode is unintuitive, so I added the following enhancement:
```vim
function! IsCommentaryOpFunc()
    return &operatorfunc ==? matchstr(maparg('<Plug>Commentary', 'n'),
            \ '\cset op\%(erator\)\?func=\zs.\{-\}\ze<cr>')
endfunction

let g:matchup_text_obj_linewise_operators = ['d', 'y',
    \ 'g@,IsCommentaryOpFunc()', 'V']
```
Now you can specify exactly which visual modes and which `operatorfunc`s (a bit complicated to determine) you want to act line-wise.

**Comment by andymass** [comment]
> Can adding g@ and : to linewise operators have side effects?

1. If you add `g@`, then any plugin which uses `operatorfunc` will be line-wise, which might be undesirable.  That's why I added the syntax `g@,execute-string`, so you can check what `opfunc` is.  If you don't care or don't use any other plugins, it won't matter.

2. I don't think adding `:` (or `v`, `V`, and `<c-v>`) will have any side-effects but I'm not completely sure.  Note, this option only applies to the `i%` text-object, so the effect should be limited.
